test/controllers/admin/layouts_controller_test.rb

```diff
@@ -0,0 +1,55 @@
+require 'test_helper'
+
+class Admin::LayoutsControllerTest < ActionDispatch::IntegrationTest
+
+  def setup
+    @grouping1 = groupings :grouping1
+    @grouping2 = groupings :grouping2
+  end
+
+  test '#edit' do
+    get edit_admin_layout_url(@grouping1)
+    assert_response :success
+  end
+
+  test '#update coordinate already exists' do
+    # the desk part of grouping2 already has a coordinate
+    assert_update(@grouping2)
+  end
+
+  test '#update coordinate does not exist' do
+    # the desks and groupings under grouping1 don't have coordinates yet
+    assert_update(@grouping1)
+  end
+
+  private
+
+  def assert_update(grouping)
+    desk_json = coordinate_json(grouping.desks)
+    grouping_json = coordinate_json(grouping.child_groupings)
+    # generate mock layout param from client for updating coordinates
+    params = { layout: { 'desk' => desk_json, 'grouping' => grouping_json }.to_json }
```

Comment:
Nit: For better readability I'd stay away from the rocket notation, i.e `'foo' => 'bar'` and stick with symbolized keys, i.e `foo: bar` and encode to json right where you make the request, in:
```ruby
put admin_layout_url(grouping), params: params.to_json
```

Reply:
I changed away from rocket notation, that was silly on my part.

For the json encoding, I want the desk to be assigned to a json string but not all params should be a json string. So the two places to call `#to_json` have different effects.